base/utf.jl

```diff
@@ -0,0 +1,880 @@
+        elseif ch < 0xe0
+            buf[out += 1] = @get_utf8_2!(str, pos, ch)
+        # Handle range 0x800-0xffff
+        elseif ch < 0xf0
+            buf[out += 1] = @get_utf8_3!(str, pos, ch)
+        # Handle range 0x10000-0x10ffff
+        else
+            ch = @get_utf8_4!(str, pos, ch)
+            @output_utf16_surr!(buf, out, ch)
+        end
+    end
+    UTF16String(buf)
+end
+
+#=
+@doc """
+@brief      Reencodes a UTF-16 or UTF-32 encoded string using UTF-8 encoding
+
+@param[in]  str::Union(Vector{UInt16}, Vector{UInt32})
+
+@return     ::UTF8String
+@throws     ArgumentError
+""" ->
+=#
+function convert(::Type{UTF8String}, str::Vector{UInt16})
+    len = sizeof(str)
+    # handle zero length string quickly
+    len == 0 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf16(str, len>>>1)
+    cnts.flags == 0 && return UTF8String(copy!(Vector{UInt8}(cnts.cntT), 1, str, 1, cnts.cntT))
+    return encode_to_8(UInt16, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function convert(::Type{UTF8String}, src::UTF16String)
+    str = src.data
+    len = sizeof(str) >>> 1
+    # handle zero length string quickly
+    len <= 1 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf16(str, len-1)
+    cnts.flags == 0 && return UTF8String(copy!(Vector{UInt8}(cnts.cntT), 1, str, 1, cnts.cntT))
+    return encode_to_8(UInt16, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function convert(::Type{UTF8String}, str::Vector{UInt32})
+    len = sizeof(str)
+    # handle zero length string quickly
+    len == 0 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf32(str, len>>>2)
+    cnts.flags == 0 && return UTF8String(copy!(Vector{UInt8}(cnts.cntT), 1, str, 1, cnts.cntT))
+    return encode_to_8(UInt32, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function convert(::Type{UTF8String},  src::UTF32String)
+    str = reinterpret(UInt32, src.data)
+    len = sizeof(str) >>> 2
+    # handle zero length string quickly
+    len <= 1 && return UTF8String("")
+    # get number of bytes to allocate
+    cnts::UTF_String_Counts = check_string_utf32(str, len-1)
+    cnts.flags == 0 && return UTF8String(copy!(Vector{UInt8}(cnts.cntT), 1, str, 1, cnts.cntT))
+    return encode_to_8(UInt32, str, cnts.cntT + cnts.cnt2 + cnts.cnt3*2 + cnts.cnt4*3)
+end
+
+function encode_to_8{T<:Union(UInt16, UInt32)}(::Type{T}, str, len)
+    buf = Vector{UInt8}(len)
+    out::UInt = 0
+    pos::UInt = 0
+    @inbounds while out < len
+        ch::UInt32 = str[pos += 1]
+        # Handle ASCII characters
+        if ch <= 0x7f
+            buf[out += 1] = ch
+        # Handle 0x80-0x7ff
+        elseif ch < 0x800
+            @output_utf8_2!(buf, out, ch)
+        # Handle 0x10000-0x10ffff (if input is UInt32)
+        elseif T == UInt32 && ch > 0xffff
```

Comment:
I don't think I understand your reply -- I don't see any function in utf16.jl which accepts either a `Vector` or and `AbstractString`.

What I'm suggesting is to force `str` to be a `Vector{T}`, so that `T` doesn't need to be provided separately. I would find it clearer (DRY principle).